### ISSUE TITLE
[no-fallthrough] Lint rule needs type information to avoid false positives e.g. nested switch

### ISSUE THREAD

**Issue by OliverJAsh** [typescript-eslint/typescript-eslint#3455]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "no-fallthrough": [2]
  }
}
```

```ts
declare const n: { a: 1; b: 1 | 2 } | { a: 2 };
(() => {
  switch (n.a) {
    case 1: {
      switch (n.b) {
        case 1:
          return 1;
        case 2:
          return 1;
      }
    }
    // ESLint error:
    // Expected a 'break' statement before 'case'.
    case 2: {
      return 2;
    }
  }
})();
```

**Expected Result**

No lint error

**Actual Result**

Lint error (see comment)

**Additional Info**

I think the TS ESLint plugin needs to provide its own version of `no-fallthrough` in order to take advantage of type information so we can avoid false positives such as this.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | latest |
| `@typescript-eslint/parser`        | latest |
| `TypeScript`                       | latest |
| `ESLint`                           | latest |
| `node`                             | latest |

**Comment by BobNobrain** [comment]
Another example of code with false positive `no-fallthrough` error due to lack of information about types:

```typescript
switch(x) {
    case 0:
        process.exit(0);
        // putting `break` here silents no-fallthrough, but causes unreachable code warning from ts

    default:
        // ...
}
```

**Comment by wclr** [comment]
Wanted to make new rule proposal but then found this. So core `no-fallthrough` requires an unneeded `break`, even if a nested switch case is exhaustive. But exhaustiveness of the case can be detected only with types. 

Such code with core `no-fallthrough enabled requires unneeded break:
```ts
const fn = (x: "A" | "B", n: "Err" | "Ok"): number => {
  switch (x) {
    case "A":
      switch (n) {
        case "Ok":
          return 1
        
        case "Err":
          return 0  
        
      }
    break; // this break actually not needed with typescript 
    default:
      return -1
  }
}
```
![image](https://user-images.githubusercontent.com/736697/152647585-06fa9c40-efa7-4bed-8819-772442a0a5c5.png)


TS config actually has the option `noFallthroughCasesInSwitch`, which almost checks the same thing, but error reporting is a bit scarce.

**Comment by dani-mp** [comment]
Had this problem today with a nested switch. The formatter removes the unnecessary `break`s. And eslint complains about them being expected.

**Comment by bradzacher** [comment]
The difficulty with building this out will be the interaction between both types and control-flow analysis.
This is definitely going to be a REALLY complicated rule for someone to build and you'll need to dig into some pretty gnarly parts of TS's APIs to get this working properly.

TS does have [`noFallthroughCasesInSwitch`](https://www.typescriptlang.org/tsconfig/#noFallthroughCasesInSwitch), though I don't know how it compares to the lint rule.

**Comment by pcone** [comment]
I don't think fixing this by changing typescript-eslint to allow the missing break statements is a good idea. This just introduces a footgun for code without 100% accurate typing, incorrect/missing data validation etc because we have no way of knowing for sure if the switch case statements will *actually* be exhaustive at runtime.

The safer and imo better fix here would be for Typescripts compile time error to allow for break statements in supposedly unreachable code somehow.

**Comment by kirkwaiblinger** [comment]
> The safer and imo better fix here would be for Typescripts compile time error to allow for break statements in supposedly unreachable code somehow.

Honestly not a crazy use of a TS error suppression 
```ts
switch (foo) {
  case bar: 
    neverReturningFunction()
    // @ts-expect-error -- ensure this branch is syntactically guaranteed to terminate
    break;
}
```

**Comment by kirkwaiblinger** [comment]
Coming back to this, and summarizing some questions raised here:

There are two separate control flow models. Eslint core's no-fallthrough can only operate based on what it sees _syntactically_, whereas TS can infer control flow based on types as well. TS's [noFallthroughCasesInSwitch](https://www.typescriptlang.org/tsconfig/#noFallthroughCasesInSwitch) option uses the TS control flow model. 

Both of these will report a violation on this [snippet](https://typescript-eslint.io/play/#ts=5.7.2&showAST=scope&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3lAXPFArgLYBGIMA3AFBVoDuAlhmABbwAUUAlPAN5XxEUNAgAMOAYMSo0SCCAB0EJAHN2AIlQh1XaoLDCEARglTp6OYuVr1GOkh3VJAemfwFH%2BFQC%2BQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaAMwEN54AXAC2n1gHNr0oB3U6XScMAXxF5A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg2RGwAYgENUqPAAtMsdgHMJAYWEl0JAJJEAygHdkeaBNYduvWqNSxtAVSKZ0wg8IBGGBbAAm6VgDNRK3gC%2BIAFAA&tokens=false)

```ts
declare const a: number;

switch (a) {
  case 0:
    console.log("one");
  case 1:
    console.log("two");

  // ... 
}
```

But TS's noFallthroughCasesInSwitch allows [this snippet](https://typescript-eslint.io/play/#ts=5.7.2&showAST=scope&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3lAXPFArgLYBGIMA3AFCiSwIBmBKYGAlqvBgBYxIDuaABQBKPChAA3ctSpp%2BbDGG7whUEfADeVeIihoEABhw7diVGiQQQAOghIA5kIBEqEM5HUzXXgOGfTMH0EAEYTb2R0K1t7J2cMfiQPWV0AelT4Gyz4KgBfIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaAMwEN54AXAC2n1gHNr0oB3U6XScMAXxF5A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg2RGwAYgENUqPAAtMsdgHMJAYWEl0JAJJEAygHdkeaBNYduvWqNSxtAVSKZ0wg8IBGGBbAAm6VgDNRK3gC%2BIAFAA&tokens=false), whereas eslint's rule does not (using a `never`-returning function call as a simpler example of type-based control flow termination than the nested `switch` in the issue example):

```ts
declare const a: number;
declare function throws(): never;

switch (a) {
  case 0:
    console.log("one");
    throws();
  case 1:
    console.log("two");

  // ... 
}
```

And, that would be sort-of ok - just use eslint no-fallthrough and put some technically-unnecessary `break`/`throw`s in:

```ts
declare const a: number;
declare function throws(): never;

switch (a) {
  case 0:
    console.log("one");
    throws();
    throw new Error('not really reachable; just satisfying eslint no-fallthrough');
  case 1:
    console.log("two");

  // ... 
}
```

However, [this code will error](https://typescript-eslint.io/play/#ts=5.7.2&showAST=scope&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3lAXPFArgLYBGIMA3AFCiSwIBmBKYGAlqvBgBYxIDuaABQBKPChAA3ctSpp%2BbDGG7whUEfADeVeIihoEABhw7diVGiQQQAOghIA5kIBEqEM5HUzXXgOGfTXR4%2BfnwQUIBRGD4YIQByFCQsOCgICABPeBTlKBJrCngAKwJMeDQodjQGdLYUB3gQNAharESAWgZUiGCkAgduOIDdMH0EAEYTb2R0K1t7J2cMfiQPWV0AenX4Gx34KgBfIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaAMwEN54AXAC2n1gHNr0oB3U6XScMAXxF5A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg2RGwAYgENUqPAAtMsdgHMJAYWEl0JAJJEAygHdkeaBNYduvWqNSxtAVSKZ0wg8IBGGBbAAm6VgDNRK3gC%2BIAFAA&tokens=false) if [allowUnreachableCode](https://www.typescriptlang.org/tsconfig/#allowUnreachableCode) is set to `false`, and, even if it's left unset, people enabling `"source.fixAll": true` in vscode (rather than the `"source.fixAll.eslint"` field) will often have the editor delete the "unneeded" statement automatically on save (see https://github.com/microsoft/TypeScript/pull/60897).

Furthermore, a lint rule, rather than a TS compiler check, is preferable because one can more easily disable/ignore it for intentional fallthroughs, whereas `@ts-expect-error` is a much more blunt instrument, which may also hide real issues.

The combination of a _lint rule_ preventing fallthroughs whose control flow detection matches TS's control flow analysis is what's proposed in this issue and what does not currently exist. 

---

As for whether TS's `allowUnreachableCode` should allow unreachable-according-to-the-types code, that would need to be proposed to TS directly, and could be used in conjunction with no-fallthrough... But, if you want that, maybe consider just using eslint core [no-unreachable](https://eslint.org/docs/latest/rules/no-unreachable) instead? 

Either way, having a no-fallthrough based on the TS control flow analysis makes sense, even if it's a justifiable judgement call to prefer the eslint core analysis instead.

**Comment by amitbeck** [comment]
So for now, I assume `no-fallthrough` should be disabled in favor of `"noFallthroughCasesInSwitch": true`?
